File: C#/0953. Verifying an Alien Dictionary.cs
public class Solution {
    public bool IsAlienSorted(string[] words, string order) {
        Dictionary<char,int> orderDict=new Dictionary<char,int>();
        for(int i=0;i<order.Length;i++){
            orderDict[order[i]]=i;
        }
        if(words.Length<=1){
            return true;
        }
        
        for(int i=1;i<words.Length;i++){
            if(!IsWordAlienSorted(words[i-1],words[i],orderDict)){
                return false;
            }
        }
        return true;
    }
    public bool IsWordAlienSorted(string wordA,string wordB,Dictionary<char,int> orderDict){
        for(int i=0;i<Math.Min(wordA.Length,wordB.Length);i++){
            if(orderDict[wordA[i]]>orderDict[wordB[i]]){
                return false;
            }
            else if(orderDict[wordA[i]]<orderDict[wordB[i]]){
                return true;
            }
        }
        if(wordA.Length<=wordB.Length){
            return true;
        }
        else{
            return false;
        }
    }
}

File: 20200215Python-China/0953. Verifying an Alien Dictionary.py
'''
In an alien language, surprisingly they also use english lowercase letters, but possibly in a different order. The order of the alphabet is some permutation of lowercase letters.

Given a sequence of words written in the alien language, and the order of the alphabet, return true if and only if the given words are sorted lexicographicaly in this alien language.

 

Example 1:

Input: words = ["hello","leetcode"], order = "hlabcdefgijkmnopqrstuvwxyz"
Output: true
Explanation: As 'h' comes before 'l' in this language, then the sequence is sorted.
Example 2:

Input: words = ["word","world","row"], order = "worldabcefghijkmnpqstuvxyz"
Output: false
Explanation: As 'd' comes after 'l' in this language, then words[0] > words[1], hence the sequence is unsorted.
Example 3:

Input: words = ["apple","app"], order = "abcdefghijklmnopqrstuvwxyz"
Output: false
Explanation: The first three characters "app" match, and the second string is shorter (in size.) According to lexicographical rules "apple" > "app", because 'l' > '∅', where '∅' is defined as the blank character which is less than any other character (More info).
 

Constraints:

1 <= words.length <= 100
1 <= words[i].length <= 20
order.length == 26
All characters in words[i] and order are English lowercase letters.
'''
class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        dict_order={}
        for k in range(len(order)):
            dict_order[order[k]]=k
        
        for k in range(1,len(words)):
            if not self.cmpOrder(words[k-1],words[k],dict_order):
                return False
        return True

    def cmpOrder(self,wordsA,wordsB,dict_order):
        if wordsA==wordsB:
            return True
        for k in range(min(len(wordsA),len(wordsB))):
            if dict_order[wordsA[k]]>dict_order[wordsB[k]]:
                return False
            elif dict_order[wordsA[k]]<dict_order[wordsB[k]]:
                return True
        if len(wordsB)<len(wordsA):
            return False
        return True
